### DISCUSSION TITLE
Any support for TF-Mini (S)?

### DISCUSSION THREAD

**Discussion by BlackHawk1942** [arendst/Tasmota#10543] [Q&A] [answered]
Hi @all!

I tried tasmota-sensors with a lidar range sensor ( [Benewake TF-Mini](https://cdn-reichelt.de/documents/datenblatt/C150/TFMINI-SA01PRODUCTMANUAL.pdf) ). Unfortunately there seems to be no way to read the data. Somewhere in the internet I read that this sensor is using the VL53L0X chip. Therefor I cloned tasmota and and wrote "#define USE_VL53L0X" into the user_config_override. Everything fine so war. The driver was loaded. As described in the manual the start address is 0x10. So I changed the 0x29 of the VL53L0X to 0x10 in the xsns_45_vl53l0x.ino.

Wireing seems ok over GPOI 4/5 --> D1/2 on an ESP8266 NodeMCU and an external 5V supply. 

Is there a native support or do I need to make other changes? Checks the device via Arduino Sketch. Data is coming and correct. I2CScanner over Arduino finds the senor without issues. I2CScan over tasmota fails.

Greetz and many thanks!

**Comment by barbudor** [comment, marked as answer]
An I2C address of 0x10 do not match with VL53L0X, neither the documented protocol. Check the datasheet.

You may want to try the serial interface of the TFmini.
Scripting may be able to process the data, or just forward to an host for processing.

**Reply by BlackHawk1942** [reply]
I2C address of 0x10 is documented in the datasheet of the TF Mini. Thats why I changed the address. TOP 5.5 "I2C data communication" as default address.

**Reply by barbudor** [reply]
It doesn't change the point that is not a valid address for the VL53L0X. That should have triggered some question before you changed the driver.
Same the I2C protocol of the TFMini do not match the VL53L0X documentation.
Before making assumptions, RTFM

So serial seems the way to go unless you want to write an I2C driver for it.

Beside the TFMini is powered by 5V. You need to insert voltage adapters between the TFMini and an ESP otherwise lifetime of the ESP will be drastically reduced.

**Reply by barbudor** [reply]
Note that also you need to change configuration between serial and I2C
How did you tested the device in Arduino ? Over serial or I2C ?

**Reply by BlackHawk1942** [reply]
> Beside the TFMini is powered by 5V. You need to insert voltage adapters between the TFMini and an ESP otherwise lifetime of the ESP will be drastically reduced.

The sensor is powerd by an external 5V supply. Ground PIN is connected with the GRD of the external 5V supply and the GRD of the sensor. 5V is used from the external supply. I don´t understand why this should effect the ESP.

> So serial seems the way to go unless you want to write an I2C driver for it.

You are talking about config as generic module and using Serial Tx / Rx on PIN Rx / Tx ? Tried that but got no data input. Even if I tried to cross Tx/Rx.

**Reply by barbudor** [reply]
You seem to lack some basic understanding of electronics o let me try to explain

When a device is powered at 5V, its I/OS are likely to be digital signals with the same range where logical "0" is close to 0.0V and logical "1" is close to its power supply.
 Unless the datasheet of the part say something different (which I've not seen on the linked doc but I've read it very quickly), you should assume that the TX signal from the TFmini will range between 0V and 5V.
 The ESP is powered by 3V3 and its GPIO are not designed to receive signals with higher amplitude than its own power. This is how you will damage the ESP in the log term, by providing a signal that reach 5V to a chip not designed for it.
 
 
Regarding the 2nd question, the TFMini is documented as supporting either serial at 115200 bps on pins TX/RX or I2C on the same pins (then labelled SCL/SDA). The doc says 
```
The product is compatible with both the UART and I2C interface. Different interfaces can be switched by commands.
``` 
(chap 2.1 page 7)
However I can't seem to find in the doc how to "switch by command".


You said that you have tested with Arduino.
Let's start from that point and elaborate what you've done
On which board ?
With which sketch ?

**Reply by barbudor** [reply]
Oh and by the way you must also connect the GND of the sensor system with the GND of the ESP system.
They must have a common reference potential.

**Reply by barbudor** [reply]
Regarding signal levels it is written:
```
Communication level LVTTL（3.3V）
```
Chap 4, table 4

So it is ok, you can connect directly to the ESP 👍

**Reply by BlackHawk1942** [reply]
Basically U R right. I`m not that familiar with Arduino and stuff. But I do know that basics. Therefore I RTFM right before I ordered and noticed that 3,3V comm-LvL :-) Otherwise I would have chosen a ardu or RPi directly ^^. GRD of the sensor and GRD of the external 5V is connected to one of the GRD PINs on the ESP. Only 5V supply is running directly to the sensor.

Due to the two interfaces on the same wires: I doesn´t have found any docu on that. I assume that this could be made in the tool provided by the manufacturer. I already have that. Unfortunately my TTL to USB converter is broken and I recently ordered a new one. Should arrive today or tomorrow.

What I have? ESP8266-12E NodeMCU V3 with CH340G Chip. 

I wanted to use I2C due to the fact that UART is less resposive. So at first I tried the normal wireing for SDA / SCL on GPOI 4/5, loaded an I2CScanner with loop and checked every data PIN - just to be sure. After get a first positive response I wanted to double checl that with a second I2CScanner. Loaded another one and got the same results. 

After that I was looking for a sketch to get some real data, no HEX. This is what I used.

`#include <Wire.h> // I2C head file`

`#define TFmini_Addr   0x10   `

`void setup() {`
 ` // put your setup code here, to run once:`
  `Serial.begin(115200);`
  `// Initiate the Wire library and join the I2C bus as a master or Slave.`
  `Wire.begin(); `
  `Serial.print("Ready to Read TFmini\r\n");`
  `delay(10);`

  `// Reset `
  `Wire.beginTransmission(TFmini_Addr);`
  `Wire.write(0x06);`
  `Wire.endTransmission(true);`

  `delay(50);`

  `// Fix detection pattern `
  `Wire.beginTransmission(TFmini_Addr);`
  `Wire.write(0x00);`
  `Wire.write(0x51);`
  `Wire.write(0x01);`
  `Wire.endTransmission(false);`
  `Wire.beginTransmission(TFmini_Addr);`
  `Wire.write(0x01);`
  `Wire.endTransmission(true);`
  `delay(50);`

  `// Set ranging mode to short distance (0.3-2m)`
 ` Wire.beginTransmission(TFmini_Addr);`
  `Wire.write(0x00);`
  `Wire.write(0x50);`
  `Wire.write(0x01);`
  `Wire.endTransmission(false);`
  `Wire.beginTransmission(TFmini_Addr);`
  `Wire.write(0x00);`
  `Wire.endTransmission(true);`

  `delay(50);  `

`}`

`void loop() `
`{`
  `// put your main code here, to run repeatedly:`
  `byte i = 0;`
  `byte rx_Num = 0;  // the bytes of received by I2C`
  `byte rx_buf[7] = {0}; // received buffer by I2C`
  
  `Wire.beginTransmission(TFmini_Addr); // Begin a transmission to the I2C Slave device with the given address.`
  `Wire.write(1); // Reg's Address_H`
  `Wire.write(2); // Reg's Address_L`
  `Wire.write(7); // Data Length`
  Wire.endTransmission(0);  // Send a START Sign`
  
  `// Wire.requestFrom（AA,BB）;receive the data form slave.`
  `// AA: Slave Address ; BB: Data Bytes `
  `rx_Num = Wire.requestFrom(TFmini_Addr, 7); `

  `// Wire.available: Retuens the number of bytes available for retrieval with read().`
 ` while( Wire.available())`
  `{`
      `rx_buf[i] = Wire.read(); // received one byte`
      `i++;`
  `}`
  
  `// OUTPUT`
  `Serial.print("TrigFlag= ");`
  `Serial.print(rx_buf[0]);`
  `Serial.print(",Dist= ");`
  `Serial.print(rx_buf[2]|(rx_buf[3] << 8));`
  `Serial.print(",Strength= ");`
  `Serial.print(rx_buf[4]|(rx_buf[5] << 8));`
  `Serial.print(",RangingMode= `");`
  `Serial.print(rx_buf[6]);`
  `Serial.print("\r\n");`

  `delay(10);`
`}`

Right after my last message here I noticed that on of my data wires wasn´t connected from sensor to the ESP. So the data was send one wired. That means this cant be real I2C - as far as I know. UNless that fact the sketch delivered data over the serial monitor but strange data. Meanwhile I think the sensors comm is UART es described.

My assumption is to rewire the data wires of the sensor to Rx/Tx of the ESP - crossed - and then start all over. But the question is: doas Tasmota give me that data?

GREETZ!

**Reply by barbudor** [reply]
My suggestion would be 1st to make it work in Serial mode with an Arduino sketch using SoftwareSerial so you can keep the original Serial of the NodeMCU to communicate with the console. Once confirmed that you have configured it in Serial mode, you could use Tasmota.

Tasmota will provide `SerialRecv` events with an hexadecimal data string of the message received.
You can process locally using Rules, or for finer control, using Scripting Language.
But the data frame will also be published over MQTT. So if you have a processing back end such as Node-Red or some Home Automation, you can set the processing there.
You can also send hexadecimal command strings using `SerialSend5  XXXXXX` 

I see you are concerned about responsiveness of the serial. Don't forget that Tasmota is not an instrumentation solution and is not meant to process multiple measurements per seconds. I would suggest to have no more than 1 measure per second.

Let us know

**Reply by BlackHawk1942** [reply]
Basically I dont not have concerns. Have see that in the past. One datapoint per minute fairly enough. And yes I do have ioBroker and NodeRed as backend system. So every logic should be done there :-)

I wil try the UART way and provide a feedback :-) Thanks for your help up to here :-)

**Reply by BlackHawk1942** [reply]
The new TTL to USB converter reached yesterday. In the manual there are different commands to change the range, protocoll and so on. I checked that and used it with the GUI-tool - gotcha! The interface was change to I2C and the Sketch provided above ist working exactly and fine on my ESP8266.

**Reply by barbudor** [reply]
So you now got the TFmini to work in I2C on the ESP.
But with Arduino sketch ? Not in Tasmota ?

I2C in Tasmota requires  a driver that needs to be written now.
While Serial can be supported directly.

**Reply by BlackHawk1942** [reply]
Yes, I2C via Adru sketch. I changed the start address of VL53L0X driver from 0x29 to 0x10. Sensor was recognized. Next step ist to rewrite ja VL53L0X and change the data read to the TFMini. Maybe I will write a new driver directly - need to have a closer look at this and getting deeper into the code.

**Reply by barbudor** [reply]
It was "recognized" because it respond too the address that you have changed
But it doesn't behave at all as the VL53L0X
As said in the doc, the I2C protocol and the Serial protocol use the same commands and responses and is totally different from the VL53L0X protocol.

So please do not try to modify the VL53L0X in the idea to make it work for TFMini but consider writing a TFMini dedicated driver
You can still use the VL53L0X driver (or any other i2C driver) as an example but don't try to merge your changes with VL53L0X.
I suggest that you develop that as XSNS_99 / XI2C_96 to avoid any clash witch other existing driver.

But I still believe that you are taking the loag road vs using Serial.

Let us know how you progress

**Reply by BlackHawk1942** [reply]
Just to point out: I am a software developer - usually for C#, JS and stuff. So I know what I shoulnd´t do ^^ Especially not merging or perform a pull req on dirty code. Since this is my first project with tasmota I wanna use the VL53L0X driver as my baseline. I soon as this driver is working (nobody knows if this will ever work^^) I will create a knew driver and maybe perfom a pull request :-)
